src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -234,32 +242,10 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
                 PatKind::Wild
             }
             // keep old code until future-compat upgraded to errors.
-            ty::Adt(adt_def, _) if !self.type_marked_structural(cv.ty) => {
-                debug!("adt_def {:?} has !type_marked_structural for cv.ty: {:?}", adt_def, cv.ty);
-                let path = tcx.def_path_str(adt_def.did);
-                let msg = format!(
-                    "to use a constant of type `{}` in a pattern, \
-                     `{}` must be annotated with `#[derive(PartialEq, Eq)]`",
-                    path, path,
-                );
-                self.saw_const_match_error.set(true);
-                tcx.sess.span_err(span, &msg);
-                PatKind::Wild
-            }
-            // keep old code until future-compat upgraded to errors.
-            ty::Ref(_, adt_ty @ ty::TyS { kind: ty::Adt(_, _), .. }, _)
-                if !self.type_marked_structural(adt_ty) =>
+            ty::Adt(adt_def, _)
+                if !self.type_marked_structural(cv.ty) && !self.backcompat_hack.get() =>
             {
-                let adt_def =
-                    if let ty::Adt(adt_def, _) = adt_ty.kind { adt_def } else { unreachable!() };
-
-                debug!(
-                    "adt_def {:?} has !type_marked_structural for adt_ty: {:?}",
-                    adt_def, adt_ty
-                );
-
-                // HACK(estebank): Side-step ICE #53708, but anything other than erroring here
-                // would be wrong. Returnging `PatKind::Wild` is not technically correct.
+                debug!("adt_def {:?} has !type_marked_structural for cv.ty: {:?}", adt_def, cv.ty);
```

Comment:
The diff here looks very weird. I'm guessing you removed the entire `ty::Ref(_, adt_ty @ ty::TyS { kind: ty::Adt(_, _), .. }, _)` arm but git has a hard time spotting that?

Can you keep the old one-line `ty::Adt(adt_def, _) if !self.type_marked_structural(cv.ty)` pattern and have a `if self.backcompat_hack.get() { return ...; }` inside?

That feels like it should fix the diff, and be more cohesive.

Reply:
I just swapped the arms, that fixed the diff, too